src/mpid/ch4/netmod/ucx/ucx_am.h

```diff
@@ -37,6 +37,59 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend(int rank,
     ucx_hdr.handler_id = handler_id;
     ucx_hdr.data_sz = data_sz;
 
+#ifdef HAS_UCP_AM_NBX
+    size_t header_size = sizeof(ucx_hdr) + am_hdr_sz;
+    void *send_buf, *header, *data_ptr;
+    /* note: since we are not copying large contig gpu data, it is less useful
+     * to use MPIR_gpu_malloc_host */
+    if (dt_contig) {
+        /* only need copy headers */
+        send_buf = MPL_malloc(header_size, MPL_MEM_OTHER);
+        MPIR_Assert(send_buf);
+        header = send_buf;
+
+        MPIR_Memcpy(header, &ucx_hdr, sizeof(ucx_hdr));
+        MPIR_Memcpy((char *) header + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
+        data_ptr = (char *) data + dt_true_lb;
+    } else {
+        /* need copy headers and pack data */
+        send_buf = MPL_malloc(header_size + data_sz, MPL_MEM_OTHER);
+        MPIR_Assert(send_buf);
+        header = send_buf;
+        data_ptr = (char *) send_buf + header_size;
+
+        MPIR_Memcpy(header, &ucx_hdr, sizeof(ucx_hdr));
+        MPIR_Memcpy((char *) header + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
+        MPI_Aint actual_pack_bytes;
+        mpi_errno = MPIR_Typerep_pack(data, count, datatype, 0, data_ptr, data_sz,
+                                      &actual_pack_bytes);
+        MPIR_ERR_CHECK(mpi_errno);
+        MPIR_Assert(actual_pack_bytes == data_sz);
```

Comment:
We could probably use the UCP generic datatype support to skip packing in the netmod and let UCX handle it. Not sure it buys us anything in practice.

Reply:
Does the UCP generic datatype use IOVs? Which we probably need to check for densities. Yeah, same with the following comment, let's leave to work on it separately.

Reply:
For generic datatypes, it just does packing, but in the UCP library via a set of function pointers we supply.